### PR TITLE
Refactoring S4 plugin

### PR DESCRIPTION
**Pull request by pinebit**
Draft.

### CODE REVIEW

File: core/services/ocr2/plugins/functions/plugin.go [hunk 1]
```diff
@@ -81,6 +86,38 @@ func NewFunctionsServices(sharedOracleArgs *libocr2.OCR2OracleArgs, conf *Functi
...
+		ORM:    s4ORM,
+		ConfigDecoder: func(configBytes []byte) (*s4.PluginConfig, *types.ReportingPluginLimits, error) {
+			wrapper, err2 := config.DecodeReportingPluginConfig(configBytes)
+			if err != nil {
+				return nil, nil, err2
+			}
+			if wrapper.Config == nil || wrapper.Config.S4PluginConfig == nil {
+				return nil, nil, errors.New("missing s4 plugin config")
+			}
+			return &s4.PluginConfig{
```
Comment (bolekk) [inline core/services/ocr2/plugins/functions/plugin.go:101]: Nice but there is a bit of a discrepancy with how TDH2 plugin handles config - it packs it into a single structure and later unpacks in NewReportingPlugin (https://github.com/smartcontractkit/tdh2/blob/main/go/ocr2/decryptionplugin/decryption.go#L36).
We should try to unify those two approaches. The one here seems to be cleaner.

**Comment by pinebit** [comment]
@bolekk I left safe changes only to merge the plugin changes alone, while waiting for other PRs.

File: core/services/ocr2/plugins/s4/plugin.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ type key struct {
 
 var _ types.ReportingPlugin = (*plugin)(nil)
 
-func NewReportingPlugin(logger logger.Logger, config *PluginConfig, orm s4.ORM) (types.ReportingPlugin, error) {
+func NewReportingPlugin(logger commontypes.Logger, config *PluginConfig, orm s4.ORM) (types.ReportingPlugin, error) {
```
Comment (bolekk) [inline core/services/ocr2/plugins/s4/plugin.go:30]: I'd suggest to rename this method to not collide with the OCR Factory interface one. Or maybe even move those validations into the factory method and get rid of this one completely?

File: core/services/ocr2/plugins/s4/plugin.go [hunk 3]
```diff
@@ -110,11 +110,11 @@ func (c *plugin) Observation(ctx context.Context, _ types.ReportTimestamp, query
 
 	queryRows, addressRange, err := UnmarshalQuery(query)
 	if err != nil {
-		c.logger.Errorw("Failed to unmarshal query (likely malformed)", "err", err)
+		c.logger.Error("Failed to unmarshal query (likely malformed)", commontypes.LogFields{"err": err})
```
Comment (bolekk) [inline core/services/ocr2/plugins/s4/plugin.go:113]: it's pretty annoying that we can't use standard Errorw() calls...

**Review by bolekk — APPROVED** [review]

File: core/services/ocr2/plugins/s4/plugin.go (hunk 2 above)
Comment (pinebit) [reply core/services/ocr2/plugins/s4/plugin.go:30]: This is convenient for tests to have this constructor function. I can rename it to NewReportingPluginInstance for example, in the next PR.

**Pull request merged**